File: stream-chat-android-docs/src/main/java/io/getstream/chat/docs/java/ui/messages/MessageList.java
package io.getstream.chat.docs.java.ui.messages;

import android.graphics.Color;
import android.text.format.DateUtils;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.concurrent.TimeUnit;

import io.getstream.chat.android.models.ReactionSortingByCount;
import io.getstream.chat.android.ui.common.helper.DateFormatter;
import io.getstream.chat.android.ui.feature.messages.list.MessageListView;
import io.getstream.chat.android.ui.feature.messages.list.adapter.BaseMessageItemViewHolder;
import io.getstream.chat.android.ui.feature.messages.list.adapter.MessageListItem;
import io.getstream.chat.android.ui.feature.messages.list.adapter.MessageListItemPayloadDiff;
import io.getstream.chat.android.ui.feature.messages.list.adapter.MessageListItemViewHolderFactory;
import io.getstream.chat.android.ui.helper.TransformStyle;
import io.getstream.chat.android.ui.viewmodel.messages.MessageListViewModel;
import io.getstream.chat.android.ui.viewmodel.messages.MessageListViewModelBinding;
import io.getstream.chat.android.ui.viewmodel.messages.MessageListViewModelFactory;
import io.getstream.chat.docs.databinding.TodayMessageListItemBinding;

/**
 * [Message List](https://getstream.io/chat/docs/sdk/android/ui/message-components/message-list/)
 */
public class MessageList extends Fragment {

    private MessageListView messageListView;

    public void usage() {
        // Init ViewModel
        ViewModelProvider.Factory factory = new MessageListViewModelFactory.Builder(requireContext())
                .cid("messaging:123")
                .build();
        ViewModelProvider provider = new ViewModelProvider(this, factory);
        MessageListViewModel viewModel = provider.get(MessageListViewModel.class);

        // Bind View and ViewModel
        MessageListViewModelBinding.bind(viewModel, messageListView, getViewLifecycleOwner());
    }

    public void handlingActions() {
        messageListView.setLastMessageReadHandler(() -> {
            // Handle when last message got read
        });
        messageListView.setEndRegionReachedHandler(() -> {
            // Handle when end region reached
        });
        messageListView.setMessageDeleteHandler((message) -> {
            // Handle when message is going to be deleted
        });
        messageListView.setThreadStartHandler((message) -> {
            // Handle when new thread for message is started
        });
        messageListView.setMessageFlagHandler((message) -> {
            // Handle when message is going to be flagged
        });
        messageListView.setMessagePinHandler((message) -> {
            // Handle when message is going to be pinned
        });
        messageListView.setMessageUnpinHandler((message) -> {
            // Handle when message is going to be unpinned
        });
        messageListView.setGiphySendHandler((giphyAction) -> {
            // Handle when some giphyAction is going to be performed
        });
        messageListView.setMessageRetryHandler((message) -> {
            // Handle when some failed message is going to be retried
        });
        messageListView.setMessageReactionHandler((message, reactionType) -> {
            // Handle when some reaction for message is going to be send
        });
        messageListView.setMessageReplyHandler((cid, message) -> {
            // Handle when message is going to be replied in the channel with cid
        });
        messageListView.setAttachmentDownloadHandler((attachmentDownloadCall) -> {
            // Handle when attachment is going to be downloaded
        });
        messageListView.setMessageEditHandler((message) -> {
            // Handle edit message
        });
    }

    public void listeners() {
        messageListView.setOnMessageClickListener((message) -> {
            // Handle message being clicked
            return true;
        });
        messageListView.setOnEnterThreadListener((message) -> {
            // Handle thread being entered
            return true;
        });
        messageListView.setOnAttachmentDownloadClickListener((attachment) -> {
            // Handle clicks on the download attachment button
            return true;
        });
        messageListView.setOnUserReactionClickListener((message, user, reaction) -> {
            // Handle clicks on a reaction left by a user
            return true;
        });
        messageListView.setOnMessageLongClickListener((message) -> {
            // Handle message being long clicked
            return true;
        });
        messageListView.setOnAttachmentClickListener((message, attachment) -> {
            // Handle attachment being clicked
            return true;
        });
        messageListView.setOnUserClickListener((user) -> {
            // Handle user avatar being clicked
            return true;
        });
    }

    public void customization() {
        TransformStyle.setMessageListItemStyleTransformer(source -> {
            // Customize the theme
            return source;
        });
        TransformStyle.setMessageListStyleTransformer(source -> {
            // Customize the theme
            return source;
        });

        /* Java requires all the fields to be provided
        TransformStyle.setMessageListItemStyleTransformer(source -> {
            return source.getReactionsViewStyle().copy(
                    source.getReactionsViewStyle().copy(
                            ReactionSortingByCount.INSTANCE
                    )
            );
        });
        */
    }

    public void channelFeatureFlags() {
        messageListView.setRepliesEnabled(false);
        messageListView.setDeleteMessageEnabled(false);
        messageListView.setEditMessageEnabled(false);
    }

    public void dateFormatter() {
        messageListView.setMessageDateFormatter(
                new DateFormatter() {
                    private final DateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy");
                    private final DateFormat timeFormat = new SimpleDateFormat("HH:mm");

                    public String formatDate(Date date) {
                        // Provide a way to format Date
                        return dateFormat.format(date);
                    }

                    public String formatTime(Date date) {
                        // Provide a way to format Time
                        return timeFormat.format(date);
                    }

                    @NonNull
                    @Override
                    public String formatRelativeTime(@Nullable Date date) {
                        // Provide a way to format relative time

                        return DateUtils.getRelativeDateTimeString(
                                getContext(),
                                date.getTime(),
                                DateUtils.MINUTE_IN_MILLIS,
                                DateUtils.WEEK_IN_MILLIS,
                                0
                                ).toString();
                    }

                    @NonNull
                    @Override
                    public String formatRelativeDate(@NonNull Date date) {
                        // Provide a way to format relative date

                        return DateUtils.getRelativeTimeSpanString(
                                date.getTime(),
                                System.currentTimeMillis(),
                                DateUtils.DAY_IN_MILLIS,
                                DateUtils.FORMAT_ABBREV_RELATIVE
                                ).toString();
                    }
                }
        );
    }

    public void customMessagesFilter() {
        String forbiddenWord = "secret";
        messageListView.setMessageListItemPredicate(item -> {
            if (item instanceof MessageListItem.MessageItem) {
                MessageListItem.MessageItem messageItem = (MessageListItem.MessageItem) item;
                return !((MessageListItem.MessageItem) item).getMessage().getText().contains(forbiddenWord);
            }

            return true;
        });
    }

    public void customMessagesView() {

        class TodayViewHolder extends BaseMessageItemViewHolder<MessageListItem.MessageItem> {

            TodayMessageListItemBinding binding;

            public TodayViewHolder(@NonNull ViewGroup parentView, @NonNull TodayMessageListItemBinding binding) {
                super(binding.getRoot());
                this.binding = binding;
            }

            @Override
            public void bindData(@NonNull MessageListItem.MessageItem data, @Nullable MessageListItemPayloadDiff diff) {
                binding.textLabel.setText(data.getMessage().getText());
            }
        }

        class CustomMessageViewHolderFactory extends MessageListItemViewHolderFactory {

            private int TODAY_VIEW_HOLDER_TYPE = 1;

            @Override
            public int getItemViewType(@NonNull MessageListItem item) {
                if (item instanceof MessageListItem.MessageItem) {
                    MessageListItem.MessageItem messageItem = (MessageListItem.MessageItem) item;
                    if (messageItem.isTheirs()
                            && messageItem.getMessage().getAttachments().isEmpty()
                            && isLessThanDayAgo((messageItem.getMessage().getCreatedAt()))) {
                        return TODAY_VIEW_HOLDER_TYPE;
                    }
                }

                return super.getItemViewType(item);

            }

            @Override
            public int getItemViewType(@NonNull BaseMessageItemViewHolder<? extends MessageListItem> viewHolder) {
                if (viewHolder instanceof TodayViewHolder) {
                    return TODAY_VIEW_HOLDER_TYPE;
                }
                return super.getItemViewType(viewHolder);
            }

            private boolean isLessThanDayAgo(Date date) {
                if (date == null) return false;
                long dayInMillis = TimeUnit.DAYS.toMillis(1);

                return date.getTime() >= System.currentTimeMillis() - dayInMillis;
            }

            @NonNull
            @Override
            public BaseMessageItemViewHolder<? extends MessageListItem> createViewHolder(@NonNull ViewGroup parentView, int viewType) {
                if (viewType == TODAY_VIEW_HOLDER_TYPE) {
                    return new TodayViewHolder(parentView, TodayMessageListItemBinding.inflate(LayoutInflater.from(parentView.getContext()), parentView, false));
                }
                return super.createViewHolder(parentView, viewType);
            }

            public void setCustomViewHolderFactory() {
                messageListView.setMessageViewHolderFactory(new CustomMessageViewHolderFactory());
            }
        }
    }

    public void customEmptyState() {
        TextView textView = new TextView(getContext());
        textView.setText("There are no messages yet");
        textView.setTextColor(Color.RED);
        messageListView.setEmptyStateView(
                textView,
                new FrameLayout.LayoutParams(
                        FrameLayout.LayoutParams.WRAP_CONTENT,
                        FrameLayout.LayoutParams.WRAP_CONTENT,
                        Gravity.CENTER
                )
        );
    }

    @SuppressWarnings({"Lambda can be replaced with method reference", "Convert2MethodRef"})
    public void avatarPredicate() {
        messageListView.setShowAvatarPredicate((messageItem) ->
                messageItem.isTheirs()
        );
    }
}

File: stream-chat-android-docs/src/main/kotlin/io/getstream/chat/docs/kotlin/ui/messages/MessageList.kt
// ktlint-disable filename

package io.getstream.chat.docs.kotlin.ui.messages

import android.graphics.Color
import android.text.format.DateUtils
import android.view.Gravity
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import io.getstream.chat.android.models.Attachment
import io.getstream.chat.android.models.Message
import io.getstream.chat.android.models.Reaction
import io.getstream.chat.android.models.ReactionSorting
import io.getstream.chat.android.models.ReactionSortingByCount
import io.getstream.chat.android.models.User
import io.getstream.chat.android.ui.common.helper.DateFormatter
import io.getstream.chat.android.ui.common.state.messages.list.GiphyAction
import io.getstream.chat.android.ui.feature.messages.list.MessageListView
import io.getstream.chat.android.ui.feature.messages.list.adapter.BaseMessageItemViewHolder
import io.getstream.chat.android.ui.feature.messages.list.adapter.MessageListItem
import io.getstream.chat.android.ui.feature.messages.list.adapter.MessageListItemPayloadDiff
import io.getstream.chat.android.ui.feature.messages.list.adapter.MessageListItemViewHolderFactory
import io.getstream.chat.android.ui.helper.StyleTransformer
import io.getstream.chat.android.ui.helper.TransformStyle
import io.getstream.chat.android.ui.viewmodel.messages.MessageListViewModel
import io.getstream.chat.android.ui.viewmodel.messages.MessageListViewModelFactory
import io.getstream.chat.android.ui.viewmodel.messages.bindView
import io.getstream.chat.docs.R
import io.getstream.chat.docs.databinding.TodayMessageListItemBinding
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.concurrent.TimeUnit

/**
 * [Message List](https://getstream.io/chat/docs/sdk/android/ui/message-components/message-list/)
 */
class MessageListViewSnippets : Fragment() {

    private lateinit var messageListView: MessageListView

    /**
     * [Usage](https://getstream.io/chat/docs/sdk/android/ui/message-components/message-list/#usage)
     */
    fun usage() {
        // Init ViewModel
        val viewModel: MessageListViewModel by viewModels {
            MessageListViewModelFactory(requireContext(), cid = "messaging:123")
        }

        // Bind View and ViewModel
        viewModel.bindView(messageListView, viewLifecycleOwner)
    }

    /**
     * [Handling Actions](https://getstream.io/chat/docs/sdk/android/ui/message-components/message-list/#handling-actions)
     */
    fun handlingActions() {
        messageListView.setLastMessageReadHandler {
            // Handle when last message got read
        }
        messageListView.setEndRegionReachedHandler {
            // Handle when end region reached
        }
        messageListView.setMessageDeleteHandler { message: Message ->
            // Handle when message is going to be deleted
        }
        messageListView.setThreadStartHandler { message: Message ->
            // Handle when new thread for message is started
        }
        messageListView.setMessageFlagHandler { message: Message ->
            // Handle when message is going to be flagged
        }
        messageListView.setMessagePinHandler { message: Message ->
            // Handle when message is going to be pinned
        }
        messageListView.setMessageUnpinHandler { message: Message ->
            // Handle when message is going to be unpinned
        }
        messageListView.setMessageMarkAsUnreadHandler() { message: Message ->
            // Handle when message is going to be marked as unread
        }
        messageListView.setGiphySendHandler { giphyAction: GiphyAction ->
            // Handle when some giphyAction is going to be performed
        }
        messageListView.setMessageRetryHandler { message: Message ->
            // Handle when some failed message is going to be retried
        }
        messageListView.setMessageReactionHandler { message: Message, reactionType: String ->
            // Handle when some reaction for message is going to be send
        }
        messageListView.setMessageReplyHandler { cid: String, message: Message ->
            // Handle when message is going to be replied in the channel with cid
        }
        messageListView.setAttachmentDownloadHandler {
            // Handle when attachment is going to be downloaded
        }
        messageListView.setMessageEditHandler { message ->
            // Handle edit message
        }
    }

    /**
     * [Listeners](https://getstream.io/chat/docs/sdk/android/ui/message-components/message-list/#listeners)
     */
    fun listeners() {
        messageListView.setOnMessageClickListener { message: Message ->
            // Handle message being clicked
            true
        }
        messageListView.setOnEnterThreadListener { message: Message ->
            // Handle thread being entered
            true
        }
        messageListView.setOnAttachmentDownloadClickListener { attachment: Attachment ->
            // Handle clicks on the download attachment button
            true
        }
        messageListView.setOnUserReactionClickListener { message: Message, user: User, reaction: Reaction ->
            // Handle clicks on a reaction left by a user
            true
        }
        messageListView.setOnMessageLongClickListener { message ->
            // Handle message being long clicked
            true
        }
        messageListView.setOnAttachmentClickListener { message, attachment ->
            // Handle attachment being clicked
            true
        }
        messageListView.setOnUserClickListener { user ->
            // Handle user avatar being clicked
            true
        }
    }

    /**
     * [Customization](https://getstream.io/chat/docs/sdk/android/ui/message-components/message-list/#customization)
     */
    fun customization() {
        TransformStyle.messageListItemStyleTransformer = StyleTransformer { defaultViewStyle ->
            defaultViewStyle.copy(
                messageBackgroundColorMine = Color.parseColor("#70AF74"),
                messageBackgroundColorTheirs = Color.WHITE,
                textStyleMine = defaultViewStyle.textStyleMine.copy(color = Color.WHITE),
                textStyleTheirs = defaultViewStyle.textStyleTheirs.copy(color = Color.BLACK),
            )
        }

        TransformStyle.messageListStyleTransformer = StyleTransformer { defaultViewStyle ->
            defaultViewStyle.copy(
                scrollButtonViewStyle = defaultViewStyle.scrollButtonViewStyle.copy(
                    scrollButtonColor = Color.RED,
                    scrollButtonUnreadEnabled = false,
                    scrollButtonIcon = ContextCompat.getDrawable(requireContext(), R.drawable.stream_ui_ic_clock)!!,
                ),
            )
        }

        TransformStyle.messageListItemStyleTransformer = StyleTransformer { defaultViewStyle ->
            defaultViewStyle.copy(
                reactionsViewStyle = defaultViewStyle.reactionsViewStyle.copy(
                    reactionSorting = ReactionSortingByCount,
                ),
            )
        }
    }

    fun channelFeatureFlags() {
        messageListView.setRepliesEnabled(false)
        messageListView.setDeleteMessageEnabled(false)
        messageListView.setEditMessageEnabled(false)
    }

    fun dateFormatter() {
        messageListView.setMessageDateFormatter(
            object : DateFormatter {
                private val dateFormat: DateFormat = SimpleDateFormat("dd/MM/yyyy")
                private val timeFormat: DateFormat = SimpleDateFormat("HH:mm")

                override fun formatDate(date: Date?): String {
                    // Provide a way to format Date
                    return dateFormat.format(date)
                }

                override fun formatTime(date: Date?): String {
                    // Provide a way to format Time
                    return timeFormat.format(date)
                }

                override fun formatRelativeTime(date: Date?): String {
                    // Provide a way to format Relative Time
                    date ?: return ""

                    return DateUtils.getRelativeDateTimeString(
                        context,
                        date.time,
                        DateUtils.MINUTE_IN_MILLIS,
                        DateUtils.WEEK_IN_MILLIS,
                        0,
                    ).toString()
                }

                override fun formatRelativeDate(date: Date): String {
                    // Provide a way to format Relative Date
                    return DateUtils.getRelativeTimeSpanString(
                        date.time,
                        System.currentTimeMillis(),
                        DateUtils.DAY_IN_MILLIS,
                        DateUtils.FORMAT_ABBREV_RELATIVE,
                    ).toString()
                }
            }
        )
    }

    fun customMessagesFilter() {
        val forbiddenWord = "secret"
        val predicate = MessageListView.MessageListItemPredicate { item ->
            !(item is MessageListItem.MessageItem && item.message.text.contains(forbiddenWord))
        }
        messageListView.setMessageListItemPredicate(predicate)
    }

    fun customMessagesView() {
        class TodayViewHolder(
            parentView: ViewGroup,
            private val binding: TodayMessageListItemBinding = TodayMessageListItemBinding.inflate(LayoutInflater.from(
                parentView.context),
                parentView,
                false),
        ) : BaseMessageItemViewHolder<MessageListItem.MessageItem>(binding.root) {

            override fun bindData(data: MessageListItem.MessageItem, diff: MessageListItemPayloadDiff) {
                binding.textLabel.text = data.message.text
            }
        }

        class CustomMessageViewHolderFactory : MessageListItemViewHolderFactory() {
            override fun getItemViewType(item: MessageListItem): Int {
                return if (item is MessageListItem.MessageItem &&
                    item.isTheirs &&
                    item.message.attachments.isEmpty() &&
                    item.message.createdAt.isLessThenDayAgo()
                ) {
                    TODAY_VIEW_HOLDER_TYPE
                } else {
                    super.getItemViewType(item)
                }
            }

            override fun getItemViewType(viewHolder: BaseMessageItemViewHolder<out MessageListItem>): Int {
                if (viewHolder is TodayViewHolder) {
                    return TODAY_VIEW_HOLDER_TYPE
                }
                return super.getItemViewType(viewHolder)
            }

            private fun Date?.isLessThenDayAgo(): Boolean {
                if (this == null) {
                    return false
                }
                val dayInMillis = TimeUnit.DAYS.toMillis(1)
                return time >= System.currentTimeMillis() - dayInMillis
            }

            override fun createViewHolder(
                parentView: ViewGroup,
                viewType: Int,
            ): BaseMessageItemViewHolder<out MessageListItem> {
                return if (viewType == TODAY_VIEW_HOLDER_TYPE) {
                    TodayViewHolder(parentView)
                } else {
                    super.createViewHolder(parentView, viewType)
                }
            }

            private val TODAY_VIEW_HOLDER_TYPE = 1
        }

        fun setCustomViewHolderFactory() {
            messageListView.setMessageViewHolderFactory(CustomMessageViewHolderFactory())
        }
    }

    fun customEmptyState() {
        val textView = TextView(context).apply {
            text = "There are no messages yet"
            setTextColor(Color.RED)
        }
        messageListView.setEmptyStateView(
            view = textView,
            layoutParams = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT,
                Gravity.CENTER
            )
        )
    }

    fun avatarPredicate() {
        messageListView.setShowAvatarPredicate { messageItem ->
            messageItem.isTheirs
        }
    }
}
